ls1intum/Artemis
# Enhancement of the simplification of the local setup/ Fails If In Production functionality

### Checklist

### Motivation and Context


The functionality of the simplification of the local setup for example submission and result simulation should be only used for local testing. To ensure that developers don't misuse these methods we introduce the method _failsIfInProduction_, which will throw an exception and show an alert that highlights that this functionality should be not used in the production environment.

### Description

The method _failsIfInProduction_ checks if the current environment is the production environment. If this is the case an alert and an exception will thrown. 

### Steps for Testing


On the test server(**Please test this with all roles: admin, instructor, tutor, student**):
1. Log in to Artemis
2. Navigate to Course Management
3. Create a programming exercise
4. Navigate to the Course Overview
5. Select the course with the created programming exercise
6. Open the programming exercise details page
=> nothing should happen, no exception in the development console or alert should be shown

Local testing:

0. Checkout this branch 
1. Start Artemis in your IDE
2. Log in to Artemis
3. Navigate to Course Management
4. Create a programming exercise without a connection to the vcs and ci 
-> normal creation procedure as for a normal programming exercise, but before you click on generate tick the checkbox "Setup without connection to the VCS and CI"

![Screenshot from 2020-04-30 18-23-16](https://user-images.githubusercontent.com/12066497/80735094-795b8500-8b10-11ea-8e3c-1fdcb8968453.png) 

5. Navigate to the Course Overview
6. Select the course with the created programming exercise
7. Open the programming exercise details page
-> nothing should happen, no exception in the development console or alert should be shown
-> Two buttons (submission and result) should be visible
![Screenshot from 2020-04-30 18-27-55](https://user-images.githubusercontent.com/12066497/80735084-72347700-8b10-11ea-968d-0344957284b4.png)
8. If click on the submission and then on the result button an alert should be shown
![Screenshot from 2020-05-09 20-15-35](https://user-images.githubusercontent.com/12066497/81481665-f33aef00-9231-11ea-9f41-105490498473.png)
![Screenshot from 2020-05-09 20-15-43](https://user-images.githubusercontent.com/12066497/81481667-f59d4900-9231-11ea-8840-37d9a21d13d1.png)

**In order to see the alert and the exception in the development console, change the following line in the `failsIfInProduction` method**
![Screenshot from 2020-05-09 20-18-35](https://user-images.githubusercontent.com/12066497/81481788-c5a27580-9232-11ea-923d-757f8425ae02.png)
to
![Screenshot from 2020-05-09 20-23-09](https://user-images.githubusercontent.com/12066497/81481809-f08cc980-9232-11ea-886a-4fcfcdb1c397.png)
-> Open the exercise details page of the created programming exercise without the connection to the vcs and ci server
-> Click on the submission button 
=> Now an alert should appear and in the development console an exception

### Screenshots



This exception and alert will be shown, if the functionality is misused in production.
![Screenshot from 2020-05-09 19-45-37](https://user-images.githubusercontent.com/12066497/81481271-17e19780-922f-11ea-8f4a-fc8986219062.png)
![Screenshot from 2020-05-09 19-45-57](https://user-images.githubusercontent.com/12066497/81481281-26c84a00-922f-11ea-9f96-9bd0074ba0ec.png)
